### PR TITLE
docker: 19.03.4 -> 20.10.2

### PR DESCRIPTION
**Pull request by mikroskeem**
###### Motivation for this change

This PR updates Docker to 20.10.2, and refactors the build process to handle Docker engine and CLI split.

Also adds long waited CGroups v2 support.

Depends on https://github.com/NixOS/nixpkgs/pull/108959

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -21229,11 +21229,10 @@ in
...
 
   inherit (callPackage ../applications/virtualization/docker {})
-    docker_18_09
-    docker_19_03;
+    docker_20_10;
 
-  docker = docker_19_03;
-  docker-edge = docker_19_03;
+  docker = docker_20_10;
+  docker-edge = docker_20_10;
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:21235]: Maybe it would be better to move this to `aliases.nix` file :thinking: ?

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:21235]: This could change in the future when they start builds for 21.04.

File: pkgs/applications/virtualization/docker/default.nix [hunk 2]
```diff
@@ -64,16 +59,73 @@ rec {
...
+      src = fetchFromGitHub {
+        owner = "moby";
+        repo = "moby";
+        rev = mobyRev;
+        sha256 = mobySha256;
+      };
+
+      goPackagePath = "github.com/docker/docker";
+
+      nativeBuildInputs = [ pkgconfig go-md2man go libtool installShellFiles ];
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:77]:
```suggestion
      nativeBuildInputs = [ pkg-config go-md2man go libtool installShellFiles ];
```

File: pkgs/applications/virtualization/docker/default.nix [hunk 3]
```diff
@@ -64,16 +59,73 @@ rec {
...
+      buildInputs = [
+        makeWrapper sqlite lvm2 btrfs-progs systemd libseccomp
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:79]: makeWrapper belongs into nativeBuildInputs.

File: pkgs/applications/virtualization/docker/default.nix [hunk 4]
```diff
@@ -64,16 +59,73 @@ rec {
...
+        export AUTO_GOPATH=1
+        export DOCKER_GITCOMMIT="${rev}"
+        export VERSION="${version}"
+        ./hack/make.sh dynbinary
+        cd -
+      '';
+
+      postPatch = ''
+        patchShebangs .
+        substituteInPlace ./hack/make.sh                   --replace libsystemd-journal libsystemd
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:99]:
```suggestion
        substituteInPlace ./hack/make.sh --replace libsystemd-journal libsystemd
```

File: pkgs/applications/virtualization/docker/default.nix [hunk 5]
```diff
@@ -64,16 +59,73 @@ rec {
...
+      goPackagePath = "github.com/docker/docker";
+
+      nativeBuildInputs = [ pkgconfig go-md2man go libtool installShellFiles ];
+      buildInputs = [
+        makeWrapper sqlite lvm2 btrfs-progs systemd libseccomp
+      ];
+
+      extraPath = optionals (stdenv.isLinux) (makeBinPath [ iproute iptables e2fsprogs xz xfsprogs procps util-linux git ]);
+
+      dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:84]: Why are disabling this?

File: pkgs/applications/virtualization/docker/default.nix [hunk 6]
```diff
@@ -108,71 +152,44 @@ rec {
...
-      cd ./go/src/${goPackagePath}/components/cli
+
+      cd ./go/src/${goPackagePath}
       # Mimic AUTO_GOPATH
       mkdir -p .gopath/src/github.com/docker/
       ln -sf $PWD .gopath/src/github.com/docker/cli
       export GOPATH="$PWD/.gopath:$GOPATH"
       export GITCOMMIT="${rev}"
       export VERSION="${version}"
+      export BUILDTIME="$(date -d @0 -u +"%Y-%m-%dT%H:%M:%SZ")" # 1970-01-01T00:00:00Z
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:163]:
```suggestion
      export BUILDTIME="1970-01-01T00:00:00Z"
```

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:21235]: At least to me it seems they stopped with edge releases (can't find a reference for Docker for Linux, but it seems that what happened for Windows/macOS at least).

File: pkgs/applications/virtualization/docker/default.nix (hunk 5 above)
Comment (mikroskeem) [reply pkgs/applications/virtualization/docker/default.nix:82]: Probably unnecessary now - looking at [other Linux distribution packaging script](https://github.com/archlinux/svntogit-community/commit/ff77881c461d895472a91f4f04da43b878bf58e5#diff-37538beb61ff63edebbf735dfcf39e5d732f49183d6beb097169d971875ca422L17), then they dropped it as well :thinking:

`# don't strip binaries! A sha1 is used to check binary consistency.` - I wonder what this really means.

**Comment by ajs124** [comment]
This should close #109156

**Review by NeQuissimus — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
> This should close #109156

Should we backport this or the 19.03.9 update?

File: pkgs/applications/virtualization/docker/default.nix [hunk 7]
```diff
@@ -64,16 +59,71 @@ rec {
...
+      src = fetchFromGitHub {
+        owner = "moby";
+        repo = "moby";
+        rev = mobyRev;
+        sha256 = mobySha256;
+      };
+
+      goPackagePath = "github.com/docker/docker";
+
+      nativeBuildInputs = [ makeWrapper pkg-config go-md2man go libtool installShellFiles ];
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/docker/default.nix:77]: Please change pkgconfig in line two to fix the eval error.

Comment (mikroskeem) [reply pkgs/applications/virtualization/docker/default.nix:77]: Done.

**Review by ghost — APPROVED** [review]
Thanks a lot, this is awesome! I tested it on a GitLab CI runner and was able to run with `systemd.enableUnifiedCgroupHierarchy = true;` just fine. Also with this release, docker respects the `--iptables=false` build-time flag afaict, so this makes things easier for me as an nftables user.

File: pkgs/applications/virtualization/docker/default.nix (hunk 5 above)
Comment (petabyteboy) [reply pkgs/applications/virtualization/docker/default.nix:82]: I guess the binary might check its own hash during runtime or something? :thinking: 
Anyways, I checked that it works without this line, so can we remove it?

Comment (mikroskeem) [reply pkgs/applications/virtualization/docker/default.nix:82]: It's already removed - see latest changes. I just left conversation in unresolved stare.

**Review by vdemeester — APPROVED** [review]
Nice, LGTM 🐯

**Comment by andersk** [comment]
> * [ ]  Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))

Shouldn’t somebody have checked this before merge? This commit broke the channel-blocking `nixos.tests.docker` test: #109416.

**Comment by petabyteboy** [comment]
I'm sorry, you're right. I only did testing in a real-world scenario. Interestingly docker.passthru.tests on aarch64-linux succeeded in ofborg. I'll look into it.

File: nixos/modules/virtualisation/docker.nix [hunk 8]
```diff
@@ -155,19 +155,17 @@ in
...
         environment = proxy_env;
         serviceConfig = {
+          Type = "notify";
           ExecStart = [
             ""
             ''
               ${cfg.package}/bin/dockerd \
                 --group=docker \
-                --host=fd:// \
+                --host=unix:// \
```
Comment (roberth) [inline nixos/modules/virtualisation/docker.nix:168]: Why did you need to change this?
Socket activation is still supported, judging from the [docs](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option).

Comment (mikroskeem) [reply nixos/modules/virtualisation/docker.nix:168]: A leftover which shouldn't ever got here.

Though I had a very odd issue when I used `fd://` - Docker kept complaining about being unable to find sockets and if I'm even launching it with systemd. Now the issue appears to be vanished 🤔

Comment (roberth) [reply nixos/modules/virtualisation/docker.nix:168]: > complaining

Was it a false warning from dockerd or an actual error, like a docker client that can't connect? If it's a warning, that seems acceptable, because then we can revert https://github.com/NixOS/nixpkgs/pull/109418#event-4209070612 and have a reliable socket activation.

Comment (mikroskeem) [reply nixos/modules/virtualisation/docker.nix:168]: No, that was an error from `dockerd`, causing daemon not to start.

Comment (mikroskeem) [reply nixos/modules/virtualisation/docker.nix:168]: Unable to reproduce the issue on fresh NixOS installation. This change should be definitely reverted.

**Comment by Mic92** [comment]
This broke kubulet from kubernetes:

```
Jan 31 21:21:32 eve kubelet[30515]: F0131 21:21:32.419449   30515 server.go:265] failed to run Kubelet: misconfiguration: kubelet cgroup driver: "cgroupfs" is different from docker cgroup driver: "systemd"
```

**Comment by Mic92** [comment]
> This broke the kubelet service from kubernetes:
> 
> ```
> Jan 31 21:21:32 eve kubelet[30515]: F0131 21:21:32.419449   30515 server.go:265] failed to run Kubelet: misconfiguration: kubelet cgroup driver: "cgroupfs" is different from docker cgroup driver: "systemd"
> ```

Addressed in https://github.com/NixOS/nixpkgs/pull/111590
